# Import Libraries

In [4]:
pip install Faker

Defaulting to user installation because normal site-packages is not writeable
                                              0.0/1.7 MB ? eta -:--:--
     ----                                     0.2/1.7 MB 3.9 MB/s eta 0:00:01
     -------                                  0.3/1.7 MB 4.1 MB/s eta 0:00:01
     -----------                              0.5/1.7 MB 3.9 MB/s eta 0:00:01
     --------------                           0.6/1.7 MB 3.7 MB/s eta 0:00:01
     -----------------                        0.8/1.7 MB 3.5 MB/s eta 0:00:01
     ----------------------                   1.0/1.7 MB 3.9 MB/s eta 0:00:01
     -------------------------                1.1/1.7 MB 3.7 MB/s eta 0:00:01
     -----------------------------            1.3/1.7 MB 3.7 MB/s eta 0:00:01
     --------------------------------         1.4/1.7 MB 3.6 MB/s eta 0:00:01
     ------------------------------------     1.6/1.7 MB 3.6 MB/s eta 0:00:01
     ---------------------------------------- 1.7/1.7 MB 3.6 MB/s eta 0

  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.


In [5]:
import pandas as pd
import random
from faker import Faker
from itertools import combinations

# Create Table 1
Table 1 contains the names and IDs of 20 sport teams. In this particular implementation, I used the `Faker` package to generate fake data for the surnames of each team 

In [107]:
# create an empty dataframe
df = pd.DataFrame()

# initialize a faker generator
fake = Faker()

# Create columns 'ID' and 'Team Name' and insert values to them
df['ID'] = range(1, 21)
df['Team Name'] = ['Team ' + fake.unique.last_name() for i in range(20)]

In [109]:
df

ID       Team Name
0    1      Team Brown
1    2   Team Shepherd
2    3     Team Miller
3    4     Team Cooper
4    5   Team Browning
5    6      Team Jones
6    7    Team Watkins
7    8   Team Gonzalez
8    9  Team Patterson
9   10     Team Carter
10  11      Team Mckee
11  12     Team Morris
12  13  Team Hernandez
13  14     Team Taylor
14  15    Team Sherman
15  16       Team Rush
16  17      Team Davis
17  18   Team Sullivan
18  19     Team Foster
19  20     Team Harris

In [ ]:
# convert the dataframe to csv
df.to_csv('table1.csv')

# Create table 2

In [73]:
# create all possible matches
teams = list(range(1, 21))

matches = list(combinations(teams, 2))

In [74]:
print('Number of matches:', len(matches))
print(matches)

Number of matches: 190
[(1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (1, 12), (1, 13), (1, 14), (1, 15), (1, 16), (1, 17), (1, 18), (1, 19), (1, 20), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (2, 12), (2, 13), (2, 14), (2, 15), (2, 16), (2, 17), (2, 18), (2, 19), (2, 20), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (3, 12), (3, 13), (3, 14), (3, 15), (3, 16), (3, 17), (3, 18), (3, 19), (3, 20), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (4, 12), (4, 13), (4, 14), (4, 15), (4, 16), (4, 17), (4, 18), (4, 19), (4, 20), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 11), (5, 12), (5, 13), (5, 14), (5, 15), (5, 16), (5, 17), (5, 18), (5, 19), (5, 20), (6, 7), (6, 8), (6, 9), (6, 10), (6, 11), (6, 12), (6, 13), (6, 14), (6, 15), (6, 16), (6, 17), (6, 18), (6, 19), (6, 20), (7, 8), (7, 9), (7, 10), (7, 11), (7, 12), (7, 13), (7, 14), (7, 15), (7, 16), (7, 17), (7, 18), (7, 19), (7, 20), (8,

As can be seen from the output above, there are 190 matches in total. Since we used combination, we are already sure that all teams have already played against every other team **at least once**. Now, we only need to add 10 more matches.

In [75]:
# Add 10 more matches by just simply switching the teams in the first 10 matches
matches += [(team2, team1) for (team1, team2) in matches[:10]]

In [76]:
print('Number of matches:', len(matches))
print(matches)

Number of matches: 200
[(1, 2), (1, 3), (1, 4), (1, 5), (1, 6), (1, 7), (1, 8), (1, 9), (1, 10), (1, 11), (1, 12), (1, 13), (1, 14), (1, 15), (1, 16), (1, 17), (1, 18), (1, 19), (1, 20), (2, 3), (2, 4), (2, 5), (2, 6), (2, 7), (2, 8), (2, 9), (2, 10), (2, 11), (2, 12), (2, 13), (2, 14), (2, 15), (2, 16), (2, 17), (2, 18), (2, 19), (2, 20), (3, 4), (3, 5), (3, 6), (3, 7), (3, 8), (3, 9), (3, 10), (3, 11), (3, 12), (3, 13), (3, 14), (3, 15), (3, 16), (3, 17), (3, 18), (3, 19), (3, 20), (4, 5), (4, 6), (4, 7), (4, 8), (4, 9), (4, 10), (4, 11), (4, 12), (4, 13), (4, 14), (4, 15), (4, 16), (4, 17), (4, 18), (4, 19), (4, 20), (5, 6), (5, 7), (5, 8), (5, 9), (5, 10), (5, 11), (5, 12), (5, 13), (5, 14), (5, 15), (5, 16), (5, 17), (5, 18), (5, 19), (5, 20), (6, 7), (6, 8), (6, 9), (6, 10), (6, 11), (6, 12), (6, 13), (6, 14), (6, 15), (6, 16), (6, 17), (6, 18), (6, 19), (6, 20), (7, 8), (7, 9), (7, 10), (7, 11), (7, 12), (7, 13), (7, 14), (7, 15), (7, 16), (7, 17), (7, 18), (7, 19), (7, 20), (8,

This part is optional but as a *sanity check*, we can create a function that checks whether all teams have already played against every other team at least once.

In [79]:
def checkTeams(matches, teams):
    # add all team pairings contained in `matches` to a set
    team_pairs = set()
    for match in matches:
        team1, team2 = match
        team_pairs.add((team1, team2))
        team_pairs.add((team2, team1))

    # check whether all teams have played against every other team
    for team1 in teams:
        for team2 in teams:
            if team1 != team2 and (team1, team2) not in team_pairs:
                return False
    return True

In [81]:
if checkTeams(matches, teams):
    print("All teams have played against every other team at least once!")
else:
    print("There are at least two teams that did not play against each other")

All teams have played against every other team at least once!


At this point, we can now create the dataframe for table 2

In [100]:
winners = []
losers = []

# Iterate over the matches and randomize which teams are the winners and losers
for match in matches:
    winner, loser = random.sample(match, k=2)
    winners.append(winner)
    losers.append(loser)

# create a dataframe containing matches with winners and losers ID
df2 = pd.DataFrame({
    'Match': [i+1 for i in range(len(matches))],
    'Winner Team ID': winners,
    'Loser Team ID': losers,
})

# set 'Match' column as the index
df2.set_index('Match', inplace=True)

In [ ]:
# convert the dataframe to csv
df2.to_csv('table2.csv')